### PR TITLE
feat(taosdump): taosdump supports dumping in/out virtual table

### PR DESCRIPTION
**Pull request by YamingPei**
# Description

taosdump supports dumping in/out virtual table

# Issue(s)

- Close https://project.feishu.cn/taosdata_td/feature/detail/6690010183

# Checklist

Please check the items in the checklist if applicable.

- [x] Is the user manual updated?
- [x] Are the test cases passed and automated?
- [x] Is there no significant decrease in test coverage?

### CODE REVIEW

**Comment by YamingPei** [comment]
失败case与PR无关：
image

File: tools/taos-tools/src/taosdump.c [hunk 1]
```diff
@@ -522,23 +523,164 @@ char * replaceNewName(char* cmd, int len) {
...
+}
+
+static inline void skip_spaces(const char *s, int n, int *i) {
+    while (*i < n && (s[*i] == ' ' || s[*i] == '\t')) (*i)++;
+}
+
+// replace db names inside FROM/USING clauses like:
+//   FROM `olddb`.`tb`.`col`    -> FROM `newdb`.`tb`.`col`
+//   USING `olddb`.`vst` ...    -> USING `newdb`.`vst` ...
+static char* replaceDbNamesInFromUsing(const char *cmd) {
```
Comment (sheyanjie-qq) [inline tools/taos-tools/src/taosdump.c:544]: this function is too complicated

File: tools/taos-tools/src/taosdump.c [hunk 2]
```diff
@@ -8879,46 +9275,65 @@ static int dumpIn() {
...
+    // physical
+    g_isVirtualPass = false;
+    ret = dumpInProcess();
+    if (ret) {
+        return ret;
+    }
+
+    // virtual
+    g_isVirtualPass = true;
+    ret = dumpInProcess();
```
Comment (sheyanjie-qq) [inline tools/taos-tools/src/taosdump.c:9337]: we should avoid scan all files again to find virtual tables

File: tools/taos-tools/src/taosdump.c (hunk 1 above)
Comment (YamingPei) [reply tools/taos-tools/src/taosdump.c:544]: The core processing logic in the code (for backticks, spaces, valid database name validation, etc.) is necessary and non-redundant. The number of effective code lines is less than 70, and there seems to be no need to split it into multiple functions.

File: tools/taos-tools/src/taosdump.c (hunk 2 above)
Comment (YamingPei) [reply tools/taos-tools/src/taosdump.c:9337]: It only reads the supertable meta files, determines whether they are virtual, and does not process subtable files, with a controllable volume of files to be handled.

**Comment by zitsen** [comment]
## Code Review Summary

This PR adds virtual table support to taosdump using a two-pass approach (physical tables first, then virtual tables). The implementation is functional but has several concerns that should be addressed.

---

## 🔴 Critical Issues

### 1. SQL Injection Risk in Query Construction

**Location:** `tools/taos-tools/src/taosdump.c` Lines 1418-1423

```c
snprintf(command, TSDB_MAX_ALLOWED_SQL_LEN,
    "SELECT COUNT(*) FROM information_schema.ins_tables "
    "WHERE stable_name = '%s' AND db_name = '%s'",
    stbName, dbName);
```

**Issue:** Direct string interpolation without escaping. If `stbName` or `dbName` contain single quotes, this could break the query or potentially cause SQL injection.

**Recommendation:** Use parameterized queries or properly escape input strings before interpolation. Similar issue exists at lines 7713-7718.

---

### 2. Memory Management in `replaceDbNamesInFromUsing()`

**Location:** Lines 545-648

**Issues:**
- Complex 100+ line function doing manual string parsing
- Multiple `malloc()`/`realloc()` calls with intricate error handling
- Caller must free returned pointer but this isn't documented
- Very difficult to test and maintain

**Recommendations:**
- Add documentation about memory ownership
- Consider breaking into smaller, testable helper functions
- Add bounds checking validation

---

### 3. Resource Cleanup in Error Paths

**Location:** Lines 6374-6397, 2633-2688

**Issue:** Several functions have error paths that return early without properly cleaning up all allocated resources.

**Example in `hasVirtualTag()`:**
```c
avro_file_reader_t reader = NULL;
if (avro_file_reader(avroFile, &reader)) {
    debugPrint("hasVirtualTag: open %s failed: %s\n", avroFile, avro_strerror());
    return false;  // Early return without cleanup
}
```

**Recommendation:** Audit all error paths to ensure proper resource cleanup (file handles, Avro readers, allocated memory).

---

## ⚠️ Major Issues

### 4. Global State Usage

**Issue:** Uses global `g_isVirtualPass` flag to control two-pass behavior. This makes the code:
- Harder to test in isolation
- Prone to state-related bugs
- Not thread-safe

**Recommendation:** Consider passing this as a parameter through the call chain instead of relying on global mutable state.

---

### 5. Magic Strings Should Be Constants

**Location:** Multiple locations

```c
const char suffix[] = "VIRTUAL 1";           // Line 2020
const char *pre_vtb = "CREATE VTABLE";       // Line 4794
```

**Recommendation:** Define these as file-scope or global constants for consistency and maintainability:

```c
#define VIRTUAL_TABLE_SUFFIX "VIRTUAL 1"
#define CREATE_VTABLE_PREFIX "CREATE VTABLE"
```

---

### 6. Insufficient Error Checking

**Location:** Line 4476-4519 in `dumpInAvroTbTagsImpl()`

```c
if (0 != avro_value_get_null(&sql_branch)) {
    // Process non-null value
}
```

**Issue:** The return value semantics of `avro_value_get_null()` aren't immediately clear. Add comments explaining what `0` vs `!= 0` means in this context.

---

## ⚙️ Code Quality Issues

### 7. Complex String Parsing Logic

The `replaceDbNamesInFromUsing()` function manually parses SQL strings character by character. While it appears to work, this approach is:
- Error-prone
- Hard to extend for new SQL patterns
- Difficult to verify correctness

**Recommendation:** Consider using a lightweight SQL tokenizer or at minimum, add comprehensive unit tests for edge cases (escaped backticks, nested quotes, etc.).

---

### 8. Missing Documentation

**Issues:**
- The two-pass algorithm (physical → virtual) isn't documented in comments
- The purpose and logic of `hasVirtualTag()` needs explanation
- Memory ownership in string manipulation functions is unclear

**Recommendation:** Add function-level comments explaining:
- Purpose and approach
- Parameter ownership semantics
- Return value meanings
- Example usage

---

## ✅ Positive Aspects

1. **Good test coverage** - The Python test cases cover the main scenarios well
2. **Backward compatibility** - Physical tables continue to work as before
3. **Logical separation** - Two-pass approach makes sense for the problem domain
4. **Defensive programming** - Added `#ifndef` guards in `toolsdef.h`
5. **Error messages** - Informative debug and error messages throughout

---

## 📋 Recommendations

1. **High Priority:**
   - Fix SQL injection vulnerability with proper escaping
   - Audit and fix resource cleanup in error paths
   - Document memory ownership in string functions

2. **Medium Priority:**
   - Refactor `replaceDbNamesInFromUsing()` into smaller functions
   - Replace global `g_isVirtualPass` with parameter passing
   - Convert magic strings to named constants

3. **Low Priority:**
   - Add unit tests for complex string parsing logic
   - Improve code comments and documentation
   - Consider extracting SQL parsing to a utility module

---

## Security Considerations

- **SQL Injection:** Critical vulnerability in query construction
- **Buffer Overflow:** Manual string manipulation is risky, though current bounds checks appear adequate
- **Resource Leaks:** Potential for file handle and memory leaks in error paths

---

Overall, the implementation achieves its goal of adding virtual table support, but needs refinement in error handling, security, and code maintainability before production deployment.
